### DISCUSSION TITLE
How to handle physical buttons

### DISCUSSION THREAD

**Discussion by Gerben1701** [PlummersSoftwareLLC/NightDriverStrip#617] [Q&A] [answered]
Greetings to all who read me,

How do I add a phyiscal button, wired to a GPIO pin and use this to cycle through effects?
Cycling through effects like the ir remote up/down buttons do.
Or by extension selecting a specific effect - though I realize this may be a bit more complicated.

Searching through the code and searching through other discussions has left me empty handed so far, some notes:
*/ In the code I so far found following useful hints:
globals.h : button pins are defined, e.g. #define TOGGLE_BUTTON_1 37
main.cpp : initialize (if button pins are defined), e.g. Bounce2::Button Button1;
remotecontrol.cpp : pressing up/down button triggers next/previous effect with method effectManager.NextEffect()
*/ In the discussions:
Zacharandir was wondering the same some time ago in [this comment from discussion #118](https://github.com/PlummersSoftwareLLC/NightDriverStrip/discussions/118#discussioncomment-1978641).
Wizard Dave directs to gitInfoPage in [discussion #8](https://github.com/PlummersSoftwareLLC/NightDriverStrip/discussions/8), but may be outdated.
Interesting, but not containing what I'm looking for [discussion #545](https://github.com/PlummersSoftwareLLC/NightDriverStrip/discussions/545).

I think this shows I did my homework.
I think this also shows I'm not classically trained in the arts of software engineering.
I hope someone may nonetheless guide me in the right direction or provide an example - if you are that person, many thanks!

**Comment by Gerben1701** [comment]
OK, after some pondering, I realized I was neglecting [Bounce2 library](https://github.com/thomasfredericks/Bounce2).

I'm starting with latching buttons.
Adding first some test code in the main.cpp loop void.
If it works, I'll have to properly initialize in the setup void, of course.

Will test soon and post results.
Feeling like a beginner, but maybe it'll help someone... like someone looking to use this NightDriver for an instructable-type project.
So for now, all you experienced programmers can ignore this discussion and focus on the interesting ones. 🙂

**Comment by Gerben1701** [comment]
Running in DEMO mode, having two effects in the effectslist (effects.cpp).
Enabled button 1 on pin 5 (globals.h).
Added code below into loop (main.cpp).
Button working just fine!

HOWEVER the auto-cycling throught the effects is in place and I'm not getting it disabled (yet).
I've tried changing the DEFAULT_EFFECT_INTERVAL value to 0 ms or 5 days, both in globals.h under DEMO and the ndef at the bottom of the file - to no effect, still auto-cycling effects.
Also I've tried disabling not-wanted effects in the loop (main.cpp), but still auto-cycling happening...
Looked at [discussion How to stop the effects auto cycling #88](https://github.com/PlummersSoftwareLLC/NightDriverStrip/discussions/88), but not applicable.

So, to be continued. In the meantime, pointers welcome! 🙂


```
        #if TOGGLE_BUTTON_1
            // UPDATE THE BUTTON BY CALLING .update() AT THE BEGINNING OF THE LOOP:
            Button1.update();

            // If the button press-state changed in this loop
            if (Button1.changed())
            {
                if (Button1.isPressed()) // Button is pressed
                {
                    g_ptrSystem->EffectManager().EnableEffect(0, false); //test
                    g_ptrSystem->EffectManager().SetCurrentEffectIndex(0);
                    //todo after successful test: insert for-loop to disable all other effects
                    g_ptrSystem->EffectManager().DisableEffect(1, false); //test
                }
                else // Button is not pressed, i.e. released
                {
                    g_ptrSystem->EffectManager().EnableEffect(1, false); // test
                    g_ptrSystem->EffectManager().SetCurrentEffectIndex(1);
                    g_ptrSystem->EffectManager().DisableEffect(0, false); // test
                }
            }
        #endif

````

**Comment by Gerben1701** [comment, marked as answer]
So setting the DEFAULT_EFFECT_INTERVAL in globals.h doesn't set this value. I'm expecting it has to do with the order of things... at some point the SetInterval method of the EffectsManager is called, may be before setting the parameter... need to look into this more.
Anyway, re-setting it in the setup loop did the trick! No more auto-cycling through effects. Calling it a workaround and calling it a day (for now).
Button is working also, all is well in the kingdom.
As stated, running in DEMO mode, but main thing is having at least two effects in the effectslist (effects.cpp).
Enabled button 1 on pin 5 (globals.h).
Impression of my code below.

So me asking a question turned into some dear-diary-moments.
Maybe it'll help someone else (another beginner), but I leave it up to the admins to decide to keep or delete this thread.
Anyway, loving the NightDriverStrip 🙂 Thanks Dave and Rutger and everybody contributing!


```
void setup()
{
...
    // ----- start custom code -----

    // Re-set default effect interval // test 2024-03-21
    g_ptrSystem->EffectManager().SetInterval(DEFAULT_EFFECT_INTERVAL, true);
    
    // UPDATE THE BUTTON BY CALLING .update()
    Button1.update();

    if (Button1.isPressed())
    {
        // Button is pressed
        g_ptrSystem->EffectManager().SetCurrentEffectIndex(1);
    }
    else
    {
        // Button is not pressed, i.e. released
        g_ptrSystem->EffectManager().SetCurrentEffectIndex(0);
    }
    
    // ------ end custom code ------
}

void loop()
{
    while(true)
    {
...
        // ----- start custom code -----

        // Set overall brightness - TODO
        // g_ptrSystem->DeviceConfig().SetBrightness(127);

        #if TOGGLE_BUTTON_1
            // UPDATE THE BUTTON BY CALLING .update() AT THE BEGINNING OF THE LOOP:
            Button1.update();

            // If the button press-state changed in this loop
            if (Button1.changed())
            {
                if (Button1.isPressed()) // Button is pressed
                {
                    g_ptrSystem->EffectManager().SetCurrentEffectIndex(1);
                }
                else // Button is not pressed, i.e. released
                {
                    g_ptrSystem->EffectManager().SetCurrentEffectIndex(0);
                }
            }
        #endif

        // ------ end custom code ------
...
    }
}
```

**Reply by Gerben1701** [reply]
_dear-diary-moments_... should've gone for captain's log...

**Comment by robertlipe** [comment]
You encouraged us to ignore you, but I'll admit I was reading enough that
I'd have nudged you back onto course if you were wandering off into the
weeds. You seemed to be making good progress on your own, so I was content
to just let you go as it seemed more educational this way.

setup() in arduino land is only run on system boot. Unless you're looking
at button presses when the system starts (which is a thing) that button
code is probably dead.

I think you've learned the value of a reasoned hack and IMO, a custom
one-off lighting project based around code like NightDriver is a good place
for those.

Trying to extend that to be useful in the general case so that buttons
could be triggered from the JSON API and state reflect in the Web UI and
having an arbitrary number of effects, including wraparound at the ends,
and the hundred other things to actually product-ize something like this
enough to check in would be non-trivial.

You didn't need a lot of computer science-y programming skills. You needed
time and determination ... and the source code of a project that _almost_
did what you wanted. THIS is the value of open source and I wish more
people would take advantage of that.

Congrats for leveling up.

Fun Fact: Most ESP32 boards have a "BOOT" button attached to GPIO0. If
you've hosed flash enough that you can't download new code, holding that
down at boot forces the bootloader to stay in the bootloader, never jumping
to user code, which allows you to reflash it.  USUALLY, that button is
needed when you first flash a new chip (or when it's done on the assembly
line) and is otherwise just a button on the board. [citation needed] After
the first half second or so after the board is powered on, that's just a
plain ole button attached to GPIO that you're free to connect to your
interrupt/polling logic and use. If you attach that to, say, a toggle
switch or otherwise have it grounded when the device reboots, be prepared
for the device to NOT boot into your code, though. But if you're
desperate for GPIOs or want to save $.05 on another button, it can be a
cool trick to remember.

That button is also often about 5mm from a reset button, so fat-fingering
it can sometimes be bad...

[Citation]:
https://docs.espressif.com/projects/esptool/en/latest/esp32/advanced-topics/boot-mode-selection.html

**Reply by Gerben1701** [reply]
Hello Robert,
Yes, it was educational! And good to know big brother was watching 😄👍
Thanks for the tip on Boot button! I'm positive this will come in handy.
On the button check in setup(), I've not tested this yet. Background: I want to use flip switches, so code needs to check after boot (setup()) what position they're in. But maybe it's caught by the first button.changed... Will test and hopefully remember to post findings.
First things first though... the kids caught me playing around with LED strips, so now I got a rainbow lighting project added to my backlog - highest priority, naturally.

**Comment by robertlipe** [comment]
Comment it out and see if the check at runtime is enough. Maybe I'm wrong.

The interaction with kids and nerdy things seems like a feature, not a bug.
Working as intended. 😉 Seriously, if they learn something and everyone has
fun, that's great.

**Comment by davepl** [comment]
Look for the definitions of TOGGLE_BUTTON.  Once they’re set, it will automatically step effects when you press that GPIO
